sources/mongodb/helpers.py

```diff
@@ -71,19 +56,93 @@ def _sort_op(self) -> List[Optional[Tuple[str, int]]]:
+        """Build a filtering operator.
+
+        Includes a field and the filtering condition for it.
+
+        Returns:
+            Dict[str, Any]: A dictionary with the filter operator.
+        """
+        if not (self.incremental and self.last_value):
+            return {}
+
+        filt = {}
+        if self.incremental.last_value_func is max:
+            filt = {self.cursor_field: {"$gte": self.last_value}}
+            if self.incremental.end_value:
+                filt[self.cursor_field]["$lt"] = self.incremental.end_value
+
+        elif self.incremental.last_value_func is min:
+            filt = {self.cursor_field: {"$lte": self.last_value}}
+            if self.incremental.end_value:
+                filt[self.cursor_field]["$gt"] = self.incremental.end_value
+
+        return filt
+
+    def _limit(self, cursor: Cursor, limit: Optional[int] = None) -> Cursor:  # type: ignore
+        """Apply a limit to the cursor, if needed.
+
+        Args:
+            cursor (Cursor): The cursor to apply the limit.
+            limit (Optional[int]): The number of documents to load.
+
+        Returns:
+            Cursor: The cursor with the limit applied (if given).
+        """
+        if (limit or 0) != 0:
+            if self.incremental is None or self.incremental.last_value_func is None:
+                logger.warning(
+                    "Using limit without ordering - results may be inconsistent."
+                )
+
+            cursor = cursor.limit(abs(limit))
+
+        return cursor
+
+    def load_documents(self, limit: Optional[int] = None) -> Iterator[TDataItem]:
+        """Construct the query and load the documents from the collection.
+
+        Args:
+            limit (Optional[int]): The number of documents to load.
+
+        Yields:
+            Iterator[TDataItem]: An iterator of the loaded documents.
+        """
+        cursor = self.collection.find(self._filter_op)
+        if self._sort_op:
+            cursor = cursor.sort(self._sort_op)
+
+        cursor = self._limit(cursor, limit)
+
+        while docs_slice := list(islice(cursor, self.chunk_size)):
+            yield map_nested_in_place(convert_mongo_objs, docs_slice)
+
+
+class CollectionLoaderParallel(CollectionLoader):
     def _get_document_count(self) -> int:
         return self.collection.count_documents(filter=self._filter_op)
 
-    def _create_batches(self) -> List[Dict[str, int]]:
+    def _create_batches(self, limit: int) -> List[Dict[str, int]]:
         doc_count = self._get_document_count()
-        return [
-            dict(skip=sk, limit=CHUNK_SIZE) for sk in range(0, doc_count, CHUNK_SIZE)
-        ]
+        if limit:
+            doc_count = min(doc_count, limit)
+
+        batches = []
+        left_to_load = doc_count
+
+        for sk in range(0, doc_count, self.chunk_size):
+            batches.append(dict(skip=sk, limit=min(self.chunk_size, left_to_load)))
+            left_to_load -= self.chunk_size
```

Comment:
As there may be situations when `limit % chink_size !=0`, we need to count how many records we're about to return, not to return more than `limit`